File: Data-Structure/sort_aglorihm/quick.cpp
#include <iostream>
using namespace std;

void swap(int *a, int *b) {
	int t = *a;
	*a = *b;
	*b = t;
}

int  partition(int arr[], int low, int high) {
	int pivot = arr[high];
	int i = low - 1;
	for (int j = low; j <= high-1; j++) {
		if (arr[j] <= pivot) {
			i++;
			swap(&arr[i], &arr[j]);
		}
	}
	swap(&arr[i+1], &arr[high]);
	return i+1;
}
void quick_sort(int arr[], int low, int high) {
	if (low < high) {
		int pivot = partition(arr, low, high);
		quick_sort(arr, low, pivot - 1);
		quick_sort(arr, pivot + 1, high);
	}
}

File: Data-Structure/sort_aglorihm/quick.js
function partition(arr, low, high) {
  let pivot = arr[high];
  let i = low - 1;
  for (let j = low; j <= high - 1; j++) {
    if (arr[j] <= pivot) {
      i++;
      [arr[i], arr[j]] = [arr[j], arr[i]];
    }
  }
  [arr[i + 1], arr[high]] = [arr[high], arr[i + 1]];
  return i + 1;
}

function quickSort(arr, low, high) {
  if (low < high) {
    let pivot = partition(arr, low, high);
    quickSort(arr, low, pivot - 1);
    quickSort(arr, pivot + 1, high);
  }
}

function finKthLargest(arr, k) {
  let l = 0;
  let h = arr.length - 1;
  k = arr.length - k;
  while (l < h) {
    let pivot  = partition(arr, l, h);
    if (pivot < k) {
      l = pivot + 1;
    } else if (pivot > k) {
      h = pivot - 1;
    } else {
      break;
    }
  }
  return arr[k];
}
var arr = [1,5,7,3,4,8];
quickSort(arr, 0, 5);
console.log(arr);

console.log(finKthLargest(arr, 2))